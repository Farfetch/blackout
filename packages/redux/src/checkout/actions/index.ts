/**
 * Checkout actions.
 */

export { default as charge } from './charge';
export { default as createCheckout } from './createCheckout';
export { default as fetchCharge } from './fetchCharge';
export { default as fetchCheckout } from './fetchCheckout';
export { default as fetchCheckoutDetails } from './fetchCheckoutDetails';
export { default as fetchCheckoutOrderOperations } from './fetchCheckoutOrderOperations';
export { default as fetchCollectPoints } from './fetchCollectPoints';
export { default as fetchDeliveryBundleUpgrades } from './fetchDeliveryBundleUpgrades';
export { default as fetchItemDeliveryProvisioning } from './fetchItemDeliveryProvisioning';
export { default as fetchUpgradeItemDeliveryProvisioning } from './fetchUpgradeItemDeliveryProvisioning';
export { default as removeCheckoutOrderItem } from './removeCheckoutOrderItem';
export { default as resetChargesState } from './resetChargesState';
export { default as resetCheckoutState } from './resetCheckoutState';
export { default as setItemTags } from './setItemTags';
export { default as setPromocode } from './setPromocode';
export { default as setTags } from './setTags';
export { default as updateCheckout } from './updateCheckout';
export { default as updateCheckoutOrderItem } from './updateCheckoutOrderItem';
export { default as updateDeliveryBundleUpgrades } from './updateDeliveryBundleUpgrades';
export { default as updateGiftMessage } from './updateGiftMessage';
