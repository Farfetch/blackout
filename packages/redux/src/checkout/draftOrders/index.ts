export * as draftOrdersActionTypes from './actionTypes.js';

export * from './actions.js';
export * from './selectors.js';

export {
  default as draftOrdersReducer,
  entitiesMapper as draftOrdersEntitiesMapper,
} from './reducer.js';

export * from './types/index.js';
