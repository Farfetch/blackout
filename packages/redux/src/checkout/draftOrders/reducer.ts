import * as actionDraftOrdersTypes from './types/draftOrdersAction.types.js';
import * as actionTypes from './actionTypes.js';
import { type AnyAction, combineReducers, type Reducer } from 'redux';

import createReducer, {
  createReducerWithResult,
} from '../../helpers/reducerFactory.js';
import type { DraftOrdersState } from './types/index.js';
import type { StoreState } from '../../types/storeState.types.js';

export const INITIAL_STATE: DraftOrdersState = {
  draftOrder: {},
  allDraftOrders: {},
  draftOrderCreations: {},
  updateDraftOrder: {},
  removeDraftOrder: {},
};

export const entitiesMapper = {
  [actionTypes.RESET_DRAFT_ORDERS_STATE]: (
    state: NonNullable<StoreState['entities']>,
  ) => {
    const { draftOrders, ...rest } = state;

    return rest;
  },
};

const draftOrderReducer = createReducer(
  'FETCH_DRAFT_ORDER',
  { isLoading: false, error: null },
  actionTypes,
);

const draftOrder = (state = INITIAL_STATE.draftOrder, action: AnyAction) => {
  if (actionDraftOrdersTypes.DRAFT_ORDER_ACTIONS.includes(action.type)) {
    return {
      ...state,
      [action.meta.draftOrderId]: draftOrderReducer(
        state[action.meta.draftOrderId],
        action,
      ),
    };
  }

  return state;
};

const allDraftOrdersReducer = createReducerWithResult(
  'FETCH_DRAFT_ORDERS',
  { isLoading: false, error: null, result: [] },
  actionTypes,
  true,
);
const allDraftOrders = (
  state = INITIAL_STATE.allDraftOrders,
  action: AnyAction,
) => {
  if (actionDraftOrdersTypes.DRAFT_ORDERS_ACTIONS.includes(action.type)) {
    return {
      ...state,
      [action.meta.hash]: allDraftOrdersReducer(
        state[action.meta.hash],
        action,
      ),
    };
  }

  return state;
};

const draftOrderCreationsReducer = createReducer(
  'CREATE_DRAFT_ORDER',
  { isLoading: false, error: null },
  actionTypes,
);
const draftOrderCreations = (
  state = INITIAL_STATE.draftOrderCreations,
  action: AnyAction,
) => {
  if (actionDraftOrdersTypes.CREATE_DRAFT_ORDER_ACTIONS.includes(action.type)) {
    return {
      ...state,
      [action.meta.orderId]: draftOrderCreationsReducer(
        state[action.meta.orderId],
        action,
      ),
    };
  }

  return state;
};

const updateDraftOrderReducer = createReducer(
  ['UPDATE_DRAFT_ORDER', 'UPDATE_DRAFT_ORDER_ITEM'],
  { isLoading: false, error: null },
  actionTypes,
);

const updateDraftOrder = (
  state = INITIAL_STATE.updateDraftOrder,
  action: AnyAction,
) => {
  if (actionDraftOrdersTypes.UPDATE_DRAFT_ORDER_ACTIONS.includes(action.type)) {
    return {
      ...state,
      [action.meta.draftOrderId]: updateDraftOrderReducer(
        state[action.meta.draftOrderId],
        action,
      ),
    };
  }

  return state;
};

const removeDraftOrderReducer = createReducer(
  ['REMOVE_DRAFT_ORDER', 'REMOVE_DRAFT_ORDER_ITEM'],
  { isLoading: false, error: null },
  actionTypes,
);

const removeDraftOrder = (
  state = INITIAL_STATE.removeDraftOrder,
  action: AnyAction,
) => {
  if (
    actionDraftOrdersTypes.REMOVE_DRAFT_ORDERS_ACTIONS.includes(action.type)
  ) {
    return {
      ...state,
      [action.meta.draftOrderId]: removeDraftOrderReducer(
        state[action.meta.draftOrderId],
        action,
      ),
    };
  }

  return state;
};

export const getDraftOrder = (state: DraftOrdersState) => state.draftOrder;
export const getAllDraftOrders = (state: DraftOrdersState) =>
  state.allDraftOrders;
export const getDraftOrderCreations = (state: DraftOrdersState) =>
  state.draftOrderCreations;
export const getUpdateDraftOrder = (state: DraftOrdersState) =>
  state.updateDraftOrder;
export const getRemoveDraftOrder = (state: DraftOrdersState) =>
  state.removeDraftOrder;

const reducer: Reducer<DraftOrdersState> = combineReducers({
  draftOrder,
  allDraftOrders,
  draftOrderCreations,
  updateDraftOrder,
  removeDraftOrder,
});

/**
 * Reducer for draft orders.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const draftOrdersReducer: Reducer<DraftOrdersState> = (state, action) => {
  if (action.type === actionTypes.RESET_DRAFT_ORDERS_STATE) {
    return reducer(INITIAL_STATE, action);
  }

  return reducer(state, action);
};

export default draftOrdersReducer;
