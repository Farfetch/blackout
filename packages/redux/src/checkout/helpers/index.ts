export { default as isCheckoutOrderAwaitingPayment } from './isCheckoutOrderAwaitingPayment.js';
export { default as isCheckoutOrderConfirmed } from './isCheckoutOrderConfirmed.js';
