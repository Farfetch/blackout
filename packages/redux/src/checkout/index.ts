export * as checkoutActionTypes from './actionTypes.js';

export * from './actions/index.js';
export * from './actions/factories/index.js';
export * from './selectors.js';

export {
  default as checkoutReducer,
  entitiesMapper as checkoutEntitiesMapper,
} from './reducer.js';

export {
  isCheckoutOrderAwaitingPayment,
  isCheckoutOrderConfirmed,
} from './helpers/index.js';

export * from './types/index.js';
export * from './draftOrders/index.js';
