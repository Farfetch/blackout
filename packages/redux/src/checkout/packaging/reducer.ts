import * as actionTypes from './../actionTypes.js';
import { type AnyAction, combineReducers } from 'redux';
import { createReducerWithResult } from '../../helpers/reducerFactory.js';
import type { PackagingOptionsState } from './index.js';

export const INITIAL_STATE: PackagingOptionsState = {
  error: null,
  result: null,
  isLoading: false,
};

export const packagingOptions = createReducerWithResult(
  'FETCH_PACKAGING_OPTIONS',
  INITIAL_STATE,
  actionTypes,
);

const error = (
  state = INITIAL_STATE.error,
  action: AnyAction,
): PackagingOptionsState['error'] => {
  switch (action.type) {
    case actionTypes.FETCH_PACKAGING_OPTIONS_FAILURE:
      return action.payload.error;
    case actionTypes.FETCH_PACKAGING_OPTIONS_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: AnyAction,
): PackagingOptionsState['isLoading'] => {
  switch (action.type) {
    case actionTypes.FETCH_PACKAGING_OPTIONS_REQUEST:
      return true;
    case actionTypes.FETCH_PACKAGING_OPTIONS_SUCCESS:
    case actionTypes.FETCH_PACKAGING_OPTIONS_FAILURE:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const result = (
  state = INITIAL_STATE.result,
  action: AnyAction,
): PackagingOptionsState['result'] => {
  switch (action.type) {
    case actionTypes.FETCH_PACKAGING_OPTIONS_SUCCESS:
      return action.payload.result;
    default:
      return state;
  }
};

export const getPackagingOptions = (
  state: PackagingOptionsState,
): PackagingOptionsState => state;

const packagingOptionsReducer = combineReducers({
  result,
  error,
  isLoading,
});

export default packagingOptionsReducer;
