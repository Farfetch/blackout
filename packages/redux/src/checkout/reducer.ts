/**
 * @module checkout/reducer
 * @category Checkout
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createMergedObject, reducerFactory } from '../helpers';
import { LOGOUT_SUCCESS } from '../authentication/actionTypes';
import assignWith from 'lodash/assignWith';
import get from 'lodash/get';
import mergeWith from 'lodash/mergeWith';
import type {
  ChargeFailureAction,
  ChargeRequestAction,
  ChargeSuccessAction,
  CompletePaymentCheckoutSuccessAction,
  DeliveryBundleFailureAction,
  DeliveryBundleRequestAction,
  FetchChargesFailureAction,
  FetchChargesRequestAction,
  FetchChargesSuccessAction,
  FetchCheckoutDetailsSuccessAction,
  FetchCollectPointsSuccessAction,
  FetchDeliveryBundleSuccessAction,
  FetchItemDeliveryProvisioningSuccessAction,
  FetchUpgradeItemDeliveryProvisioningSuccessAction,
  GenericCheckoutAction,
  GenericCheckoutFailureAction,
  GenericCheckoutRequestAction,
  GenericCheckoutSuccessAction,
  LogoutAction,
  ResetChargesStateAction,
  ResetCheckoutStateAction,
  State,
  StateWithResult,
  UpdateDeliveryBundleSuccessAction,
} from './types';
import type { StoreState } from '../types';

export const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  completePaymentCheckout: {
    error: null,
    isLoading: false,
  },
  checkoutDetails: {
    error: null,
    isLoading: false,
  },
  collectPoints: {
    error: null,
    isLoading: false,
  },
  itemTags: {
    error: null,
    isLoading: false,
  },
  promoCode: {
    error: null,
    isLoading: false,
  },
  tags: {
    error: null,
    isLoading: false,
  },
  giftMessage: {
    error: null,
    isLoading: false,
  },
  charges: {
    error: null,
    result: null,
    isLoading: false,
  },
  deliveryBundleUpgrades: {
    error: null,
    isLoading: false,
    result: null,
  },
  itemDeliveryProvisioning: {
    error: null,
    isLoading: false,
  },
  upgradeItemDeliveryProvisioning: {
    error: null,
    isLoading: false,
  },
};

const error = (
  state = INITIAL_STATE.error,
  action:
    | GenericCheckoutFailureAction
    | GenericCheckoutRequestAction
    | ResetCheckoutStateAction
    | LogoutAction,
): State['error'] => {
  switch (action?.type) {
    case actionTypes.CREATE_CHECKOUT_FAILURE:
    case actionTypes.FETCH_CHECKOUT_FAILURE:
    case actionTypes.UPDATE_CHECKOUT_FAILURE:
      return action.payload.error;
    case actionTypes.CREATE_CHECKOUT_REQUEST:
    case actionTypes.FETCH_CHECKOUT_REQUEST:
    case actionTypes.UPDATE_CHECKOUT_REQUEST:
    case actionTypes.RESET_CHECKOUT_STATE:
    case LOGOUT_SUCCESS:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const id = (
  state = INITIAL_STATE.id,
  action:
    | GenericCheckoutSuccessAction
    | ResetCheckoutStateAction
    | LogoutAction
    | FetchCheckoutDetailsSuccessAction,
) => {
  switch (action?.type) {
    case actionTypes.CREATE_CHECKOUT_SUCCESS:
    case actionTypes.UPDATE_CHECKOUT_SUCCESS:
    case actionTypes.FETCH_CHECKOUT_SUCCESS:
    case actionTypes.FETCH_CHECKOUT_DETAILS_SUCCESS:
    case actionTypes.SET_ITEM_TAGS_SUCCESS:
    case actionTypes.SET_PROMOCODE_SUCCESS:
    case actionTypes.SET_TAGS_SUCCESS:
      return action.payload.result;
    case actionTypes.RESET_CHECKOUT_STATE:
    case LOGOUT_SUCCESS:
      return INITIAL_STATE.id;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: GenericCheckoutAction | ResetCheckoutStateAction | LogoutAction,
) => {
  switch (action?.type) {
    case actionTypes.CREATE_CHECKOUT_REQUEST:
    case actionTypes.FETCH_CHECKOUT_REQUEST:
    case actionTypes.UPDATE_CHECKOUT_REQUEST:
      return true;
    case actionTypes.CREATE_CHECKOUT_FAILURE:
    case actionTypes.CREATE_CHECKOUT_SUCCESS:
    case actionTypes.FETCH_CHECKOUT_FAILURE:
    case actionTypes.FETCH_CHECKOUT_SUCCESS:
    case actionTypes.UPDATE_CHECKOUT_FAILURE:
    case actionTypes.UPDATE_CHECKOUT_SUCCESS:
    case actionTypes.RESET_CHECKOUT_STATE:
    case LOGOUT_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const convertCheckoutOrder = (
  state: StoreState['entities'],
  action: GenericCheckoutSuccessAction,
): StoreState['entities'] => {
  const { entities, result } = action.payload;

  const customizer = (objValue: any, srcValue: any, key: string) => {
    if (
      Array.isArray(objValue) ||
      key === 'shippingAddress' ||
      key === 'billingAddress'
    ) {
      return srcValue;
    }
  };

  const tempMergedState = {};
  assignWith(tempMergedState, state, customizer);
  const entitiesMerged = mergeWith({}, entities, customizer);
  const mergedState = createMergedObject(tempMergedState, entitiesMerged);

  const shouldResetClickAndCollect =
    !!get(mergedState, `checkoutOrders[${result}].clickAndCollect`) &&
    !get(entities, `checkoutOrders[${result}].clickAndCollect`);

  if (!shouldResetClickAndCollect) {
    return mergedState;
  }

  return {
    ...mergedState,
    checkoutOrders: {
      ...mergedState.checkoutOrders,
      [result]: {
        collectpoints: mergedState.checkoutOrders[result].collectpoints,
        ...entities?.checkoutOrders[result],
      },
    },
  };
};

const mergeCheckoutOrder = (
  state: StoreState['entities'],
  action:
    | CompletePaymentCheckoutSuccessAction
    | FetchCollectPointsSuccessAction,
): StoreState['entities'] => {
  const { id } = action.meta;
  const currentCheckoutOrder = get(state, `checkoutOrders[${id}]`);

  return {
    ...state,
    checkoutOrders: {
      [id]: {
        ...currentCheckoutOrder,
        ...action.payload.entities,
      },
    },
  };
};

export const entitiesMapper = {
  [actionTypes.COMPLETE_PAYMENT_CHECKOUT_SUCCESS as typeof actionTypes.COMPLETE_PAYMENT_CHECKOUT_SUCCESS]:
    mergeCheckoutOrder,
  [actionTypes.FETCH_COLLECT_POINTS_SUCCESS as typeof actionTypes.FETCH_COLLECT_POINTS_SUCCESS]:
    mergeCheckoutOrder,
  [actionTypes.FETCH_CHECKOUT_DETAILS_SUCCESS as typeof actionTypes.FETCH_CHECKOUT_DETAILS_SUCCESS]:
    (
      state: StoreState['entities'],
      action: FetchCheckoutDetailsSuccessAction,
    ): StoreState['entities'] => {
      const { id } = action.meta;
      const currentCheckout = get(state, `checkout[${id}]`);
      const { entities } = action.payload;
      const mergedState = createMergedObject(state, entities);

      return {
        ...mergedState,
        checkout: { [id]: { ...currentCheckout, checkoutOrder: id } },
      };
    },
  [actionTypes.FETCH_ITEM_DELIVERY_PROVISIONING_SUCCESS as typeof actionTypes.FETCH_ITEM_DELIVERY_PROVISIONING_SUCCESS]:
    (
      state: StoreState['entities'],
      action: FetchItemDeliveryProvisioningSuccessAction,
    ): StoreState['entities'] => {
      const { deliveryBundleId } = action.meta;
      const currentDeliveryBundles = get(state, 'deliveryBundles') || {};
      const currentDeliveryBundle = get(
        state,
        `deliveryBundles[${deliveryBundleId}]`,
      );

      return {
        ...state,
        deliveryBundles: {
          ...currentDeliveryBundles,
          [deliveryBundleId]: {
            ...currentDeliveryBundle,
            ...action.payload.entities,
            itemsIds: action.payload.result,
          },
        },
      };
    },
  [actionTypes.FETCH_UPGRADE_ITEM_DELIVERY_PROVISIONING_SUCCESS as typeof actionTypes.FETCH_UPGRADE_ITEM_DELIVERY_PROVISIONING_SUCCESS]:
    (
      state: StoreState['entities'],
      action: FetchUpgradeItemDeliveryProvisioningSuccessAction,
    ): StoreState['entities'] => {
      const { deliveryBundleId, upgradeId } = action.meta;
      const currentDeliveryUpgrades = get(state, 'deliveryBundleUpgrades');
      const selectedDeliveryUpgrades = get(
        state,
        `deliveryBundleUpgrades[${deliveryBundleId}]`,
      );
      const { result: itemsId, entities } = action.payload;

      const upgradesWithItemsDeliveryProvisioing = itemsId.reduce(
        (acc: { [x: string]: any }, itemID: string) => {
          const currentItemDeliveryProvisioning = get(
            entities,
            `itemDeliveryProvisioning[${itemID}]`,
          );
          return {
            ...acc,
            [itemID]: {
              ...acc[itemID],
              provisioning: {
                upgradeId,
                ...currentItemDeliveryProvisioning.provisioning,
              },
            },
          };
        },
        selectedDeliveryUpgrades,
      );

      return {
        ...state,
        deliveryBundleUpgrades: {
          ...currentDeliveryUpgrades,
          [deliveryBundleId]: {
            ...upgradesWithItemsDeliveryProvisioing,
          },
        },
      };
    },
  [actionTypes.UPDATE_CHECKOUT_SUCCESS as typeof actionTypes.UPDATE_CHECKOUT_SUCCESS]:
    (
      state: StoreState['entities'],
      action: GenericCheckoutSuccessAction,
    ): StoreState['entities'] => convertCheckoutOrder(state, action),
  [actionTypes.CREATE_CHECKOUT_SUCCESS as typeof actionTypes.CREATE_CHECKOUT_SUCCESS]:
    (
      state: StoreState['entities'],
      action: GenericCheckoutSuccessAction,
    ): StoreState['entities'] => convertCheckoutOrder(state, action),
  [actionTypes.SET_PROMOCODE_SUCCESS as typeof actionTypes.SET_PROMOCODE_SUCCESS]:
    (
      state: StoreState['entities'],
      action: GenericCheckoutSuccessAction,
    ): StoreState['entities'] => convertCheckoutOrder(state, action),
  [actionTypes.FETCH_CHECKOUT_SUCCESS as typeof actionTypes.FETCH_CHECKOUT_SUCCESS]:
    (
      state: StoreState['entities'],
      action: GenericCheckoutSuccessAction,
    ): StoreState['entities'] => convertCheckoutOrder(state, action),
  [actionTypes.SET_ITEM_TAGS_SUCCESS as typeof actionTypes.SET_ITEM_TAGS_SUCCESS]:
    (
      state: StoreState['entities'],
      action: GenericCheckoutSuccessAction,
    ): StoreState['entities'] => convertCheckoutOrder(state, action),
  [actionTypes.SET_TAGS_SUCCESS as typeof actionTypes.SET_TAGS_SUCCESS]: (
    state: StoreState['entities'],
    action: GenericCheckoutSuccessAction,
  ): StoreState['entities'] => convertCheckoutOrder(state, action),
  [actionTypes.RESET_CHECKOUT_STATE as typeof actionTypes.RESET_CHECKOUT_STATE]:
    (state: StoreState['entities']): StoreState['entities'] => {
      const {
        checkout,
        checkoutDetails,
        checkoutOrders,
        checkoutOrderItems,
        ...rest
      } = state;

      return { ...rest };
    },
  [LOGOUT_SUCCESS as typeof LOGOUT_SUCCESS]: (
    state: StoreState['entities'],
  ): StoreState['entities'] => {
    const {
      checkout,
      checkoutDetails,
      checkoutOrders,
      checkoutOrderItems,
      ...rest
    } = state;

    return { ...rest };
  },
};

export const completePaymentCheckout = reducerFactory(
  'COMPLETE_PAYMENT_CHECKOUT',
  INITIAL_STATE.completePaymentCheckout,
  actionTypes,
  true,
);

export const checkoutDetails = reducerFactory(
  'FETCH_CHECKOUT_DETAILS',
  INITIAL_STATE.checkoutDetails,
  actionTypes,
  true,
);

export const collectPoints = reducerFactory(
  'FETCH_COLLECT_POINTS',
  INITIAL_STATE.collectPoints,
  actionTypes,
  true,
);

export const itemTags = reducerFactory(
  'SET_ITEM_TAGS',
  INITIAL_STATE.itemTags,
  actionTypes,
  true,
);

export const promoCode = reducerFactory(
  'SET_PROMOCODE',
  INITIAL_STATE.promoCode,
  actionTypes,
  true,
);

export const tags = reducerFactory(
  'SET_TAGS',
  INITIAL_STATE.tags,
  actionTypes,
  true,
);

export const giftMessage = reducerFactory(
  'UPDATE_GIFT_MESSAGE',
  INITIAL_STATE.giftMessage,
  actionTypes,
  true,
);

export const charges = (
  state = INITIAL_STATE.charges,
  action:
    | LogoutAction
    | FetchChargesSuccessAction
    | FetchChargesFailureAction
    | FetchChargesRequestAction
    | ChargeSuccessAction
    | ChargeFailureAction
    | ChargeRequestAction
    | ResetChargesStateAction,
): StateWithResult => {
  switch (action?.type) {
    case actionTypes.CHARGE_REQUEST:
    case actionTypes.FETCH_CHARGES_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.charges.error,
        isLoading: true,
      };
    case actionTypes.CHARGE_FAILURE:
    case actionTypes.FETCH_CHARGES_FAILURE:
      return {
        ...state,
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.CHARGE_SUCCESS:
    case actionTypes.FETCH_CHARGES_SUCCESS:
      return {
        error: INITIAL_STATE.charges.error,
        isLoading: false,
        result: action.payload,
      };
    case actionTypes.RESET_CHARGES_STATE:
    case LOGOUT_SUCCESS:
      return INITIAL_STATE.charges;
    default:
      return state;
  }
};

export const deliveryBundleUpgrades = (
  state = INITIAL_STATE.deliveryBundleUpgrades,
  action:
    | LogoutAction
    | DeliveryBundleRequestAction
    | UpdateDeliveryBundleSuccessAction
    | FetchDeliveryBundleSuccessAction
    | DeliveryBundleFailureAction,
): StateWithResult => {
  switch (action?.type) {
    case actionTypes.UPDATE_DELIVERY_BUNDLE_UPGRADE_REQUEST:
    case actionTypes.UPDATE_DELIVERY_BUNDLE_UPGRADES_REQUEST:
    case actionTypes.FETCH_DELIVERY_BUNDLE_UPGRADES_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.deliveryBundleUpgrades.error,
        isLoading: true,
      };
    case actionTypes.UPDATE_DELIVERY_BUNDLE_UPGRADE_SUCCESS:
    case actionTypes.UPDATE_DELIVERY_BUNDLE_UPGRADES_SUCCESS:
      return {
        ...state,
        error: INITIAL_STATE.deliveryBundleUpgrades.error,
        isLoading: false,
      };
    case actionTypes.UPDATE_DELIVERY_BUNDLE_UPGRADE_FAILURE:
    case actionTypes.UPDATE_DELIVERY_BUNDLE_UPGRADES_FAILURE:
    case actionTypes.FETCH_DELIVERY_BUNDLE_UPGRADES_FAILURE:
      return {
        ...state,
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.FETCH_DELIVERY_BUNDLE_UPGRADES_SUCCESS:
      return {
        error: INITIAL_STATE.charges.error,
        isLoading: false,
        result: action.payload.result,
      };
    case LOGOUT_SUCCESS:
      return INITIAL_STATE.deliveryBundleUpgrades;
    default:
      return state;
  }
};

export const itemDeliveryProvisioning = reducerFactory(
  'FETCH_ITEM_DELIVERY_PROVISIONING',
  INITIAL_STATE.itemDeliveryProvisioning,
  actionTypes,
  true,
);

export const upgradeItemDeliveryProvisioning = reducerFactory(
  'FETCH_UPGRADE_ITEM_DELIVERY_PROVISIONING',
  INITIAL_STATE.upgradeItemDeliveryProvisioning,
  actionTypes,
  true,
);

export const getError = (state: State): State['error'] => state.error;
export const getId = (state: State): State['id'] => state.id;
export const getIsLoading = (state: State): State['isLoading'] =>
  state.isLoading;

export const getCompletePaymentCheckout = (
  state: State,
): State['completePaymentCheckout'] => state.completePaymentCheckout;
export const getCheckoutDetails = (state: State): State['checkoutDetails'] =>
  state.checkoutDetails;
export const getCollectPoints = (state: State): State['collectPoints'] =>
  state.collectPoints;
export const getItemTags = (state: State): State['itemTags'] => state.itemTags;
export const getPromoCode = (state: State): State['promoCode'] =>
  state.promoCode;
export const getTags = (state: State): State['tags'] => state.tags;
export const getGiftMessage = (state: State): State['giftMessage'] =>
  state.giftMessage;
export const getCharges = (state: State): State['charges'] => state.charges;
export const getDeliveryBundleUpgrades = (
  state: State,
): State['deliveryBundleUpgrades'] => state.deliveryBundleUpgrades;
export const getItemDeliveryProvisioning = (
  state: State,
): State['itemDeliveryProvisioning'] => state.itemDeliveryProvisioning;
export const getUpgradeItemDeliveryProvisioning = (
  state: State,
): State['upgradeItemDeliveryProvisioning'] =>
  state.upgradeItemDeliveryProvisioning;

/**
 * Reducer for checkout state.
 *
 * @function checkoutReducer
 * @static
 * @param {object} state - Current redux state.
 * // @ts-expect-error ts-migrate(7006) FIXME: Parameter 'state' implicitly has an 'any' type.
 * @param {object} action - Action dispatched.
 *
 * // @ts-expect-error ts-migrate(7006) FIXME: Parameter 'state' implicitly has an 'any' type.
 * @returns {object} New state.
 */
export default combineReducers({
  error,
  id,
  isLoading,
  completePaymentCheckout,
  checkoutDetails,
  collectPoints,
  itemTags,
  promoCode,
  tags,
  giftMessage,
  charges,
  deliveryBundleUpgrades,
  itemDeliveryProvisioning,
  upgradeItemDeliveryProvisioning,
});
