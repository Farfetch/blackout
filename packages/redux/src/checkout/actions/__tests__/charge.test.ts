import { actionTypes } from '../..';
import { charge } from '..';
import { INITIAL_STATE } from '../../reducer';
import { mockCharges } from 'tests/__fixtures__/checkout';
import { mockStore } from '../../../../tests';
import { postCharges } from '@farfetch/blackout-client/checkout';
import find from 'lodash/find';

jest.mock('@farfetch/blackout-client/checkout', () => ({
  ...jest.requireActual('@farfetch/blackout-client/checkout'),
  postCharges: jest.fn(),
}));

describe('charge() action creator', () => {
  const checkoutMockStore = (state = {}) =>
    mockStore({ checkout: INITIAL_STATE }, state);
  const data = {};
  const orderId = 12345;
  const expectedConfig = undefined;
  let store;

  beforeEach(() => {
    jest.clearAllMocks();
    store = checkoutMockStore();
  });

  it('should create the correct actions for when the charge procedure fails', async () => {
    const expectedError = new Error('charges error');

    postCharges.mockRejectedValueOnce(expectedError);
    expect.assertions(4);

    try {
      await store.dispatch(charge(orderId, data));
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(postCharges).toHaveBeenCalledTimes(1);
      expect(postCharges).toHaveBeenCalledWith(orderId, data, expectedConfig);
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          { type: actionTypes.CHARGE_REQUEST },
          {
            type: actionTypes.CHARGE_FAILURE,
            payload: { error: expectedError },
          },
        ]),
      );
    }
  });

  it('should create the correct actions for when the charge procedure is successful', async () => {
    postCharges.mockResolvedValueOnce(mockCharges);
    await store.dispatch(charge(orderId, data));

    const actionResults = store.getActions();

    expect(postCharges).toHaveBeenCalledTimes(1);
    expect(postCharges).toHaveBeenCalledWith(orderId, data, expectedConfig);

    expect(actionResults).toMatchObject([
      { type: actionTypes.CHARGE_REQUEST },
      {
        type: actionTypes.CHARGE_SUCCESS,
        payload: mockCharges,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.CHARGE_SUCCESS,
      }),
    ).toMatchSnapshot('charge success payload');
  });
});
