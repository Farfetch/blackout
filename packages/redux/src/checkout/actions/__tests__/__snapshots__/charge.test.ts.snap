// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`charge() action creator should create the correct actions for when the charge procedure is successful: charge success payload 1`] = `
Object {
  "payload": Object {
    "cancelUrl": "some url",
    "id": "00000000-0000-0000-0000-000000000000",
    "redirectUrl": "some url",
    "returnUrl": "some url",
    "status": "Processing",
  },
  "type": "@farfetch/blackout-redux/CHARGE_SUCCESS",
}
`;
